### PR TITLE
 #162949039 Implement User Profile

### PR DESCRIPTION
**Pull request by Teatoller**
#### What does this PR do?
Enable user to view and edit their profile
#### How should this be manually tested?
1. Navigate to a folder where you would prefer to test the repository.

2. Create a directory
 ```$```  ```mkdir myproject```

3. Navigate to the directory "myproject"
 ```$```   ```cd myproject```
4. Clone the author's haven repository
  ```$```   ```git clone https://github.com/andela/ah-robotics-frontend.git```
5. Change the directory into the "ah-robotics-fronted"
  ```$```   ```cd ah-robotics-frontend```
6.  Checkout to the branch ft-user-profile-page-162949039
 ```$```   ```git checkout ft-user-profile-page-162949039```
7. Install dependancies from the package.json
 ```$```   ```npm install``` to install all the modules listed in the `package.json` file.

8. Start the application in development mode
  ```$```   ```npm start``` to run the app in the development mode 

The application opens http://localhost:3000 in the browser.
Change the URL on the browser to;
 ```
https://ah-robotics-frontend-sta-pr-13.herokuapp.com

```

The browser will show like represented in screenshot _**image: 1**_

9. Click on profile on the Navbar and enter a valid email **for demo purpose use**;

**Email:** ```steven.otieno1@gmail.com``` 
**Password:** ```test123#```

and click on the ```login``` button, see _**image: 2**_.

10. On the welcome page, as in  _**Image: 3**_, click on ```profile``` on the navbar.
The profile is displayed as shown in _**image 4**_

11. Click on the ```Edit Profile``` **button** it will lead you to _**image 5**_ to make changes.

In the Bio input text area, type in what you want to have as your **bio**, e.g qualifications, interest and the **Click** on the ```Save``` **button** once you have completed.

See _**image: 6a**_ and _**image: 6b**_ before and after click on ```Save``` **button**.

12. **Nb:** The ```Cancel``` **button** will return you to the profile page if you choose not to save changes. 

13. To run tests, open the terminal and run the following command:
See **_Image: 7_**
```
npm test
``` 
14. To run tests and coverage, open the terminal and run the following command:
```
npm test -- --coverage
```
screenshot 2019-02-21 at 07 14 45

#### What are the relevant pivotal tracker stories?
[#162949039](https://www.pivotaltracker.com/story/show/162949039)
##### Image:1 Authors Haven landing page

screenshot 2019-02-27 at 10 45 16

##### Image: 2 Login page
screenshot 2019-02-22 at 07 35 08

##### Image: 3 Welcome page _the username_  for steven.otieno1@gmail.com is _aspine_

screenshot 2019-02-27 at 10 48 50

##### Image: 4 Profile page 

screenshot 2019-02-27 at 10 51 56

##### Image: 5 Profile Edit Page.

screenshot 2019-02-22 at 07 56 19

##### Image: 6 a & b Editing profile sample.
**Image: 6a**

screenshot 2019-02-27 at 11 01 32

**image: 6b**
screenshot 2019-02-27 at 10 54 00

##### Image: 7 Test run on branch
screenshot 2019-02-21 at 07 23 11

### CODE REVIEW

File: src/components/Profile/ProfileEdit/index.jsx [hunk 1]
```diff
@@ -0,0 +1,51 @@
+import React from 'react';
+import {
+    Image, Grid, Container, Button, Form, Input, Segment, Divider,
+} from 'semantic-ui-react';
+import { NavLink } from 'react-router-dom';
+import './profile.scss';
+import PropTypes from 'prop-types';
+import imglove from '../Images/lov.jpeg';
+
+// const user = profile;
```
Comment (johnwayodi) [inline src/components/Profile/ProfileEdit/index.jsx]: Remove this line of commented code

File: src/components/resetPassword/resetpassword.scss [hunk 2]
```diff
@@ -6,11 +6,11 @@ body > div, body > div > div,body > div > div > div.login-form {
     padding-bottom: 10px;
     border-bottom: solid 1px rgba(230, 230, 230, 0.658);
 }
-.btn-login{
-    border-radius: 20px !important;
-    padding: 5px !important;
-    height: 35px;
-}
+// .btn-login{
```
Comment (johnwayodi) [inline src/components/resetPassword/resetpassword.scss:9]: Remove this commented block of code

File: src/views/UserProfileView/profileEditView/index.js [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+    const updateData = {};
+    const { bio } = this.state;
+    updateData.data = { bio };
+    // updateData.token = token;
+    updateData.username = username;
+    profileUpdateAction(updateData);
+  };
+
+  render() {
+    // console.log(this.props);
```
Comment (johnwayodi) [inline src/views/UserProfileView/profileEditView/index.js]: There is commented code on this file. Kindly remove them

**Review by johnwayodi — COMMENTED** [review]
@Teatoller the functionality works well, Nice work.

However consider changing the **color** of the buttons to conform to the website colors, the shade of green you have used on the Profile buttons is not the same as the other buttons of the application.

File: src/components/Profile/ProfileEdit/index.jsx (hunk 1 above)
Comment (Teatoller) [reply src/components/Profile/ProfileEdit/index.jsx]: Hi @johnwayodi,
Thank you for this feedback.
I have taken note and made the adjustment.

Regards
Steven Ennis

File: src/components/resetPassword/resetpassword.scss (hunk 2 above)
Comment (Teatoller) [reply src/components/resetPassword/resetpassword.scss:9]: Hi @johnwayodi,
This is not part of my original work within this scope and will need a consultation before I expunge the commented line item on the .scss file.

File: src/views/UserProfileView/profileEditView/index.js (hunk 3 above)
Comment (Teatoller) [reply src/views/UserProfileView/profileEditView/index.js]: Hi, @johnwayodi 
Thank you for your attention to detail.
I have corrected by removing the line item.

Regards
Steven Ennis

**Review by Paulvitalis200 — COMMENTED** [review]
Hey @Teatoller ,
This feature is working well. All the functionality is good. However, we can refine the styling so that the page looks more appealing.

**Comment by Teatoller** [comment]
Hello, 
We can bring our minds together and see just what we can tweak to bring about a great UI/UX experience.

> Hey @Teatoller ,
> This feature is working well. All the functionality is good. However, we can refine the styling so that the page looks more appealing.

Regards
Steven Ennis

File: src/components/Register/register.scss [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+  color: rgba(148, 144, 144, 0.829);
+  font-family: -apple-system, BlinkMacSystemFont, Roboto, sans-serif;
+  font-size: small;
+}
+
+.form-container p a {
+  color: $primary-color;
+}
+
+// .button {
```
Comment (jomasim) [inline src/components/Register/register.scss]: remove the commented code

File: src/components/Profile/index.jsx [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+          </Grid.Row>
+          <GridRow>
+            <div>
+              <Button positive><NavLink to="editprofile"><span id="edit-profile">Edit Profile</span></NavLink></Button>
+            </div>
+          </GridRow>
+          <Grid.Row>
+            <Menu compact>
+              <Menu.Item as="a">
+                <Icon name="users" />
```
Comment (jomasim) [inline src/components/Profile/index.jsx:36]: You can format this line correctly

File: src/components/Register/register.scss (hunk 4 above)
Comment (Teatoller) [reply src/components/Register/register.scss]: @jomasim ,

Thank you. I have made the change.
Regards
Steven Ennis

File: src/components/Article/article.test.js [hunk 6]
```diff
@@ -7,8 +7,5 @@ it('it should render single article view', () => {
   expect(wrapper.length).toBe(1);
   expect(wrapper.find('container')).toBeDefined();
 });
-it('it should match snapshot', () => {
```
Comment (jomasim) [inline src/components/Article/article.test.js:10]: Why did you remove this test?

File: src/components/Profile/index.jsx [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+                {' '}
+Followers
+                <Label color="green" floating>
+                  {user.profile && user.profile.followers}
+                </Label>
+              </Menu.Item>
+              <Menu.Item as="a">
+                <Icon name="users" />
+                {' '}
+Following
```
Comment (jomasim) [inline src/components/Profile/index.jsx]: You can format this line correctly

File: src/redux/reducers/userProfileReducer/userProfileReducer.test.js [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+    expect(userProfileReducer(userProfileReducer.initialState, {})).toEqual({
+        profileData: {},
+        error: {},
+        profileUpdate: false,
+        isLoading: false,
+        Followers: 0,
+        Following: 0,
+});
+  });
+  it('it should show isloading on profile submit', () => {
```
Comment (jomasim) [inline src/redux/reducers/userProfileReducer/userProfileReducer.test.js]: isLoading should be set to true instead on submit. You can check this test again.

File: src/redux/reducers/userProfileReducer/userProfileReducer.test.js [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+    expect(userProfileReducer(null, {
+      type: PROFILE_DATA,
+      payload: {},
+    })).toEqual({ isLoading: false });
+  });
+  it('it should show isloading on profile submit', () => {
```
Comment (jomasim) [inline src/redux/reducers/userProfileReducer/userProfileReducer.test.js]: This test has the same name as the one above it. What is the difference?

File: src/redux/reducers/userProfileReducer/userProfileReducer.test.js [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+      payload: {},
+    })).toEqual({ isLoading: true, errors: {} });
+  });
+  it('it should show error on profile submit', () => {
+    expect(userProfileReducer(null, {
+      type: PROFILE_PATCH_START,
+      payload: {},
+    })).toEqual({ isLoading: true, profileData: {} });
+  });
+  it('it should show error on profile submit', () => {
```
Comment (jomasim) [inline src/redux/reducers/userProfileReducer/userProfileReducer.test.js]: Type success action is not supposed to return errors as the name of the test suggests.

File: src/components/Article/article.test.js (hunk 6 above)
Comment (Teatoller) [reply src/components/Article/article.test.js:10]: Hi @jomasim,
This may have been accidental but not for any particular reason.
My apology.

Regards
Steven Ennis

File: src/App.js [hunk 11]
```diff
@@ -11,6 +11,8 @@ import VerifyRegistrationView from './views/VerifyRegistrationView/index';
 import ArticlesView from './views/ArticlesListingView';
 import CreateArticle from './views/CreateArticleView';
 import ArticleView from './views/ArticleView';
+import ProfileView from './views/UserProfileView';
+import profileEditView from './views/UserProfileView/profileEditView';
```
Comment (jomasim) [inline src/App.js]: Place all your views inside the views folder don't nest them.
```/views/UserProfileView/profileEditView';``` 
**Correct format:**
```/views/UserProfileView/```
```/views/profileEditView/```

**Comment by jomasim** [comment]
I managed to test the functionality, it works as expected maybe you will need to check your code to resolve issues I raised above. You need also to ensure to display the same profile photo on the edit page just before someone edits it. I see you are picking a different one instead.

File: src/App.js (hunk 11 above)
Comment (Teatoller) [reply src/App.js]: Hello @jomasim,

I have successfully implemented moving the folder _profileEditView_ directly into the views.
I also renamed it to conform with our folder convention to **ProfileEditView**
Thank you 
Regards
Steven Ennis

File: src/components/Profile/index.jsx (hunk 7 above)
Comment (Teatoller) [reply src/components/Profile/index.jsx]: This is now corrected.

File: src/redux/reducers/userProfileReducer/userProfileReducer.test.js (hunk 8 above)
Comment (Teatoller) [reply src/redux/reducers/userProfileReducer/userProfileReducer.test.js]: @jomasim  
Thank you for the note.
The test is okay, but not the narration which I have edited. Thank you.

File: src/redux/reducers/userProfileReducer/userProfileReducer.test.js (hunk 9 above)
Comment (Teatoller) [reply src/redux/reducers/userProfileReducer/userProfileReducer.test.js]: Thank you, I have made changes

File: src/redux/reducers/userProfileReducer/userProfileReducer.test.js (hunk 10 above)
Comment (Teatoller) [reply src/redux/reducers/userProfileReducer/userProfileReducer.test.js]: Thank you, I have made necessary adjustment

**Review by waracci — COMMENTED** [review]
Looks good to me, besides the few issues that Joma has already pointed out.

File: src/redux/actions/userProfileActions/actions.js [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+    PROFILE_DATA,
+    PROFILE_ERROR,
+    PROFILE_PATCH_START,
+    UPLOAD_IMAGE_ERROR,
+    UPLOAD_IMAGE_SUCCESS,
+    UPLOAD_IMAGE_START,
+  } from './types';
+
+    export const profileData = payload => ({
+ type: PROFILE_DATA,
```
Comment (Collins33) [inline src/redux/actions/userProfileActions/actions.js]: This indentation seems off

File: src/redux/actions/userProfileActions/userProfileActions.test.js [hunk 13]
```diff
@@ -0,0 +1,33 @@
...
+  UPLOAD_IMAGE_SUCCESS,
+  UPLOAD_IMAGE_ERROR,
+} from './types';
+import {
+profileData, Error, profileUpdate, userImageUpload,
+        userImageUploadSuccess, userImageUploadError,
+} from './actions';
+
+describe('Profile action creator', () => {
+it('should dispatch PROFILE_DATA', () => {
```
Comment (Collins33) [inline src/redux/actions/userProfileActions/userProfileActions.test.js]: The indentation on this file looks off also.

File: src/redux/sagas/LoginSaga/loginSaga.js [hunk 14]
```diff
@@ -12,8 +11,8 @@ function* loginWorker({ payload }) {
       payload: { user: response.data.user },
     });
     const { token } = response.data.user;
-    setUserToken(token);
-    setCurrentUser(response.data.user);
+    localStorage.setItem('accessToken', JSON.stringify(token));
```
Comment (Collins33) [inline src/redux/sagas/LoginSaga/loginSaga.js:14]: Why was this part changed to stop using the `setUserToken` function?

File: src/redux/sagas/UserProfileSaga/UserProfilePatchSaga.js [hunk 15]
```diff
@@ -0,0 +1,34 @@
...
+  } catch (err) {
+    yield put({
+      type: types.PROFILE_PATCH_ERROR,
+      payload: { errors: err.response },
+    });
+  }
+}
+
+function* watchuserProfilePatch() {
+  yield takeEvery(
```
Comment (Collins33) [inline src/redux/sagas/UserProfileSaga/UserProfilePatchSaga.js:29]: Why did you choose to use `takeEvery` and not `takeLatest`?

File: src/views/ProfileEditView/index.js [hunk 16]
```diff
@@ -0,0 +1,89 @@
...
+  constructor(props) {
+    super(props);
+    this.state = {
+      bio: '',
+      image: '',
+    };
+  }
+
+
+  componentWillReceiveProps(nextProps) {
```
Comment (Collins33) [inline src/views/ProfileEditView/index.js:18]: What is happening here?

File: src/views/ProfileEditView/profileEditView.test.js [hunk 17]
```diff
@@ -0,0 +1,32 @@
...
+function setup() {
+  const profile = {};
+  const handleSubmit = () => ({});
+  const onChange = () => ({});
+
+  const wrapper = shallow(
+    <ProfileEditComponent handleSubmit={handleSubmit} profile={profile} onChange={onChange} />,
+);
+
+  return {
```
Comment (Collins33) [inline src/views/ProfileEditView/profileEditView.test.js:16]: Why are you returning the wrapper?

File: src/views/UserProfileView/index.js [hunk 18]
```diff
@@ -0,0 +1,48 @@
...
+  componentDidMount() {
+    const user = JSON.parse(localStorage.getItem('user'));
+    const payload = {};
+    const { username } = user;
+    payload.username = username;
+    const { profileData: profileAction } = this.props;
+    profileAction(payload);
+  }
+
+  componentWillReceiveProps(nextProps) {
```
Comment (Collins33) [inline src/views/UserProfileView/index.js]: Why is this empty?

**Review by Collins33 — CHANGES_REQUESTED** [review]

File: src/redux/actions/userProfileActions/actions.js (hunk 12 above)
Comment (Teatoller) [reply src/redux/actions/userProfileActions/actions.js]: Hello, 
Thank you for noticing.
You are correct I have followed through and made the correction.

Regards
Steven Ennis

File: src/redux/actions/userProfileActions/userProfileActions.test.js (hunk 13 above)
Comment (Teatoller) [reply src/redux/actions/userProfileActions/userProfileActions.test.js]: Hi, It actually is.
I have reviewed and corrected.
Thank you.

Regards
Steven Ennis

File: src/redux/sagas/LoginSaga/loginSaga.js (hunk 14 above)
Comment (Teatoller) [reply src/redux/sagas/LoginSaga/loginSaga.js:14]: Hello,
This was done because a helper function was set up in utils from which the application is able to call.

Regards
Steven Ennis

File: src/redux/sagas/UserProfileSaga/UserProfilePatchSaga.js (hunk 15 above)
Comment (Teatoller) [reply src/redux/sagas/UserProfileSaga/UserProfilePatchSaga.js:29]: Hi,
I was considering the user might want to send several requests.
When this happens the application gets to deal with the latest request.
Otherwise no other particular reason.
You may advise me on which is most appropriate, thank you.

Regards
Steven Ennis

File: src/views/UserProfileView/index.js (hunk 18 above)
Comment (Teatoller) [reply src/views/UserProfileView/index.js]: Hello,
This is extinct. I found a workaround for the implementation I had intended for it.
I have expunged it from the code base.
Thank you.

File: src/views/ProfileEditView/index.js (hunk 16 above)
Comment (Teatoller) [reply src/views/ProfileEditView/index.js:18]: Hi,
The application state values are initially empty.
In order to update the state with new prop values this method 
_"componentWillReceiveProps(nextProps) "_ gets called whenever their is a change in prop values.

Regards
Steven Ennis

File: src/views/ProfileEditView/profileEditView.test.js (hunk 17 above)
Comment (Teatoller) [reply src/views/ProfileEditView/profileEditView.test.js:16]: This wrapper shrouds the "ProfileEditComponent"
It is a more elaborate way of implementing the wrapper function.
It calls
screenshot 2019-02-28 at 11 54 04


And tests that a button exists within the ProfileEditComponent.

Regards
Steven Ennis
